pip/qsharp/_qsharp.py

```diff
@@ -84,6 +84,22 @@ def callback(output):
     return get_interpreter().run(entry_expr, shots, callback)
 
 
+def run_histogram(entry_expr, shot_count, on_result=None):
```

Comment:
I'd be happy to just fold this into the `run()` method. `on_result` can be an optional named argument, and if it's passed in, then we run in "callback mode".

```python
def run(entry_expr, *shots, on_result=None):
  if on_result is None:
      def callback(output):
          print(output)
      return get_interpreter().run(entry_expr, shots, callback)
  else:
      results = []
      
      def OnEvent(output):
          results[len(results) - 1]["events"].append(output)
      
      for i in range(shot_count):
          results.append({"events": [], "result": None})
          results[i]["result"] = (get_interpreter().run(entry_expr, 1, OnEvent))[0]
          if on_result:
              on_result(results[i])
      
      return results
```

Later on, maybe the native interpreter take result callbacks directly, then we can clean this up without changing the api signature.

Reply:
Now that I think about it, I'm not sure if we even need that first `if` case.... we can just use the full implementation from `run_histogram`. I think I just didn't want the `shots` parameter on the `run()` method to go to waste :)

Reply:
I'm inclined to agree. Then native code refactoring can be a follow up.

Reply:
Yeah, that was my original inclination. I just didn't want to mess with run when experimenting with getting something working. I can do the consolidation.